sematic/api_client.py

```diff
@@ -97,15 +106,136 @@ def _put(endpoint, json_payload) -> Any:
     return response.json()
 
 
-class APIConnectionError(requests.exceptions.ConnectionError):
+class APIConnectionError(ConnectionError):
+    pass
+
+
+class IncompatibleClientError(Exception):
+    pass
+
+
+class ServerError(Exception):
     pass
 
 
+def _validate_server_compatibility(
+    tries: int = 5, seconds_between_tries: int = 10, use_cached: bool = True
+):
+    """Check that the client is compatible with the server.
+
+    Raises an error if the server and client are incompatible, or if this can't be
+    verified.
+    """
+    global _validated_client_version
+    if _validated_client_version and use_cached:
+        return
+
+    retriable_errors = (
+        ConnectionError,
+        ServerError,
+    )
+    response = None
+    error: Optional[Exception] = None
+    for _ in range(tries):
+        response, error = _validate_server_compatibility_no_catch()
+        if error is None:
+            _validated_client_version = True
+            return
+        if not isinstance(error, retriable_errors):
+            break
+        logger.warning(
+            "Unsuccessful at determining version compatibility waiting %s seconds...",
+            seconds_between_tries,
+        )
+        time.sleep(seconds_between_tries)
+    if response is not None and response.status_code != 200:
+        logger.error(
+            "Sematic server returned status code %s when"
+            " asked for version information. Response: %s",
+            response.status_code,
+            response.text,
+        )
+
+    assert error is not None  # should be impossible to fail, but makes mypy happy
+    raise error
+
+
+def _validate_server_compatibility_no_catch() -> Tuple[
+    Optional[requests.Response], Optional[Exception]
+]:
+    headers = {"Content-Type": "application/json"}
+    response = None
+    error: Optional[Exception] = None
+    try:
+        response = requests.get(_url("/meta/versions"), headers=headers)
```

Comment:
Why not use the centralized `_get` function defined in this module? Then we can have retry logic implemented there instead of ad-hoc in `_validate_server_compatibility`.

Reply:
Ah I see why now, because you are calling this is `_request`, which is used by `_get`. Still, anyway to make the retry logic more reusable? Maybe use https://pypi.org/project/retry/?